### ISSUE TITLE
eplfs - doesn't seem to adhere to the min. space 

### ISSUE THREAD

**Issue by Symbiot78** [trapexit/mergerfs#363]
Hi

I've set mergerfs to leave 30G of space left and am using eplfs..
but it seems to keep writing..
see screenshot..

![2017-01-24_06-47-34](https://cloud.githubusercontent.com/assets/3314249/22235629/1182f6b8-e201-11e6-9614-7c70419fb376.jpg)
![2017-01-24_06-48-31](https://cloud.githubusercontent.com/assets/3314249/22235657/2e42f10e-e201-11e6-84f2-465318a4ed42.jpg)

params are: defaults,allow_other,moveonenospc=true

am I missing something?

**Comment by trapexit** [comment]
I can't really respond without more information. Does the path exist on other drives? It's a path preserving policy and as such only considers drives where the path exists. What do you mean "keeps writing"? mergerfs doesn't split files across drives. minfreespace is used when calculating where to create a file. Not while writing. It doesn't concatenate drives together. It merges them.

**Comment by ekw** [comment]
I just encountered this myself and came here to find this exact issue already here.  
My params are: category.create=eplfs,defaults,allow_other,minfreespace=5G

I have two data drives, each 4GB.  One (HD1) has filled up such that there is only 300MB free, which I didn't notice until today.  The other HD (HD2) is still completely empty.  When I create a new directory and a new file to a non-existing path (/newdir/newfile1), it still puts it on HD1.

Just for my clarification, if a file /a/b/c/d/file1 exists, and I create /a/file2, is that considered an existing path and so file2 will go on the same drive as file1?  If I change parameters in fstab, do I just need to unmount and remount the pool to have them take effect?

**Comment by trapexit** [comment]
As the docs indicate the policies filter drives and if nothing is found to met the requirements for the policy it falls back to another policy. I did just notice that the minfreespace check is missing in `mfs` but that doesn't explain what you're seeing.

Can you provide me with *precisely* the setup before and after the file/directory creation? Each drive's available storage, size, directories on each (with regard to the path in question).

**Comment by trapexit** [comment]
And as for the clarification... the existing path is ... the subtree in question. /a is the subtree if you create /a/file.

**Comment by Symbiot78** [comment]
hi

I'll try and answer your questions asap.. but in the meantime... 

eplfs... is this how I should think it:

I save a file to the mergerFS pool..

mergerFS then looks for a disk or disks with the path already there.. it finds 3 disks... it then sorts them by size and selects the disk with the least space but of course with space left for the file and then writes the file.. 
I would think that somewhere in the above it checks for the min. freespace (I've set 30) and if a disk has 32Gb of free space and I'm trying to save a 10Gb file MergerFS should then skip that disk because it would exceed the min freespace requirement set by me and move on to another disk?

That's not what I am seeing since I have several disks with space WAY below the 30 Gb limit I set..

So is the above completely wrong?

**Comment by trapexit** [comment]
You're wrong in that you are looking at the problem at a higher level than happens under the covers and while it could possibly work that way it'd be far more complicated.

There is no such thing as "save a file". There is "open/create file", "write a bunch of data", "maybe sync it", "close file". There is no way to know up front how much storage a file will consume. It could be 0 bytes or 1TB. It's not atomic. Given that: mergerfs gets a request to create a file and runs the policy, gets the most appropriate drive for it to be on, clones a path if necessary from another drive, then issues the real create. If the writer of the file keeps writing past the minfreespace barrier then it's left to do so. Since it could be doing other things with the file it's difficult to say what's appropriate. The one thing that is less unknown is how to handle running out of storage. That's what `moveonenospc` is for.

tl;dr: minfreespace is evaluated and acted on at creation time because all other times are more costly and could lead to issues. If the drive continues to get *new* files after it goes below minfreespace then most certainly it's because the policy fell back to another policy and that's where it ended up.

I need particular to explain a specific situation but the docs are pretty thorough on what happens.
